kernel: Add k_thread_foreach profiling API

ramakrishnapallala opened the pull request:
Added k_thread_foreach() API
Deprecated k_call_stacks_analyze() API
Replaced k_call_stacks_analyze() with k_thread_foreach()

Commit:
kernel: Add k_thread_foreach API

Add k_thread_foreach API to iterate over all the threads in
the system.

This API can be used for debugging threads in multi threaded
environment to dump and analyze various thread parameters like
priority, state, stack address etc...

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
I will be pushing another commit which will have test related changes.

tyrantlizard77 reviewed (changes requested):
I need you to amend one of our existing thread test cases to additionally prove that these new APIs work properly.

ramakrishnapallala commented on kernel/thread.c:
```diff
@@ -36,6 +36,38 @@ extern struct _static_thread_data _static_thread_data_list_end[];
+{
+	struct k_thread *thread;
+
+	__ASSERT(analyze, "thread analyze arguments can not be NULL");
```
Yes. will fix it.

ramakrishnapallala commented:
@andrewboie updated the PR with tests. Please have a look at it.

ramakrishnapallala commented on kernel/thread.c:
```diff
@@ -36,6 +36,38 @@ extern struct _static_thread_data _static_thread_data_list_end[];
+#endif
+
+#if defined(CONFIG_THREAD_STACK_INFO) && defined(CONFIG_THREAD_MONITOR)
+void k_thread_stack_foreach(stack_analyze_fn stack_analyze, void *user_data)
```
1. it needs separate Kconfig
2. there are cases where the app/middleware is on;y interested in stack info (for example networking)

andrewboie commented on tests/kernel/profiling/profiling_api/src/test_profiling_api.c:
```diff
@@ -74,5 +75,38 @@ void test_call_stacks_analyze_workq(void)
+{
+	test_var = 0;
+	k_thread_foreach(thread_analyze, &test_var);
+	zassert_true(test_var, "thread_analyze() not getting called");
```
This doesn't prove that it completely works. It passes even if the function just runs once, and not for every thread on the system.
Can you make this stronger, test for a specific value of test_var?

andrewboie commented on tests/kernel/profiling/profiling_api/src/test_profiling_api.c:
```diff
@@ -74,5 +75,38 @@ void test_call_stacks_analyze_workq(void)
+{
+	test_var = 0;
+	k_thread_stack_foreach(stack_analyze, &test_var);
+	zassert_true(test_var, "stack_analyze() not getting called");
```
Same complaint as above.

ramakrishnapallala commented:
@pizi-nordic @andrewboie please review the updated PR.

andrewboie commented on include/kernel.h:
```diff
@@ -549,6 +552,39 @@ enum execution_context_types {
+ *
+ * @return N/A
+ */
+extern void k_thread_foreach(thread_analyze_fn thread_analyze, void *user_data);
```
think we could just call this thread_callback or something.

andrewboie commented on tests/kernel/profiling/profiling_api/src/test_profiling_api.c:
```diff
@@ -74,5 +82,96 @@ void test_call_stacks_analyze_workq(void)
+	thread_count = 0;
+	temp = 0;
+	k_thread_stack_foreach(stack_analyze, &temp);
+	zassert_true(thread_count && temp,
```
This still doesn't prove that this API works.
It passes even if the callback only fires once and not for each thread in the system.

ramakrishnapallala replied:
@andrewboie the test is done in a relative way. I have called the API twice. In the first run, thread callback function counts the no.of threads present in the system. Then a new thread is being created. Now the k_thread_foreach() is again called. This time the thread count should be more than the previous one and we are also checking the tid is matching with newly added thread.

same approach is followed for k_thread_stack_foreach().

Note: I do not want to use the _kernel.threads to compare the tid's in the test's callback.

rgundi commented on kernel/thread.c:
```diff
@@ -36,6 +36,38 @@ extern struct _static_thread_data _static_thread_data_list_end[];
 	     thread_data++)
 
+#if defined(CONFIG_THREAD_MONITOR)
+void k_thread_foreach(thread_user_cb user_cb, void *user_data)
```
The argument user_cb should be a const.

ramakrishnapallala replied:
this is not a struct data. It is function pointer and the intention is not to be const function()

rgundi commented on kernel/thread.c:
```diff
@@ -36,6 +36,38 @@ extern struct _static_thread_data _static_thread_data_list_end[];
+#endif
+
+#if defined(CONFIG_THREAD_STACK_INFO) && defined(CONFIG_THREAD_MONITOR)
+void k_thread_stack_foreach(stack_user_cb user_cb, void *user_data)
```
The argument user_cb should be a const.

ramakrishnapallala replied:
same here.

pizi-nordic commented on kernel/thread.c:
```diff
@@ -36,6 +36,38 @@ extern struct _static_thread_data _static_thread_data_list_end[];
+#endif
+
+#if defined(CONFIG_THREAD_STACK_INFO) && defined(CONFIG_THREAD_MONITOR)
+void k_thread_stack_foreach(stack_user_cb user_cb, void *user_data)
```
The k_thread_stack_foreach() still do not pass any information about inspected thread to the user_cb. As result threads with low stack usage cannot be identified.

Also, thread stack inspection might be performed using  k_thread_foreach(), which is more flexible because it gives access to whole thread structure. IMHO presence of  k_thread_foreach() makes k_thread_stack_foreach() obsolete.

ramakrishnapallala replied:
valid point.

@jukkar @nashif let me know your opinion.

lpereira commented on include/kernel.h:
```diff
@@ -533,7 +533,10 @@ enum execution_context_types {
  * @ingroup kernel_apis
  * @{
  */
+typedef void (*thread_user_cb)(const struct k_thread *thread, void *user_data);
```
Minor, but, per convention, function pointer typedefs in kernel.h all follow the `some_name_t` convention rather than `some_name_cb` one.

ramakrishnapallala replied:
OK.

lpereira commented on kernel/thread.c:
```diff
@@ -36,6 +36,38 @@ extern struct _static_thread_data _static_thread_data_list_end[];
+
+	__ASSERT(user_cb, "user_cb can not be NULL");
+
+	for (thread = _kernel.threads; thread != NULL;
```
The iteration should happen while holding a lock; otherwise, if the linked list is modified midway, it'll either be called with invalid parameters, or following the next pointer will lead you to someplace you shouldn't be going.

lpereira commented on kernel/thread.c:
```diff
@@ -36,6 +36,38 @@ extern struct _static_thread_data _static_thread_data_list_end[];
+
+	__ASSERT(user_cb, "user_cb can not be NULL");
+
+	for (thread = _kernel.threads; thread != NULL;
```
Same thing for locks here.

ramakrishnapallala replied:
agree. I will have add the lock while adding as well.

lpereira reviewed (changes requested):
While there are some issues with the patch, it would be nice to know why you need this thing in the first place.  Please amend the commit message, maybe giving examples.

ramakrishnapallala commented:
Addressed @lpereira comment's. and updated the PR.

@jukkar do we need explicit k_thread_stack_foreach() API or you can use k_thread_foreach() and get the stack details from thread structs?

nashif commented:
> @jukkar do we need explicit k_thread_stack_foreach() API or you can use k_thread_foreach() and get the stack details from thread structs?

I think we can do without k_thread_stack_foreach and let user implement this using k_thread_foreach

ramakrishnapallala commented:
@pizi-nordic @lpereira @nashif  I have updated the PR as per your comments. Please have a look.

@andrewboie  I have added a test case which tests the API in a relative manner. Please have a look a look.

lpereira commented on kernel/thread.c:
```diff
@@ -36,6 +36,26 @@ extern struct _static_thread_data _static_thread_data_list_end[];
+	while (thread != NULL) {
+		user_cb(thread, user_data);
+
+		key = irq_lock();
```
This lock won't help.  It needs to surround the whole loop.

ramakrishnapallala replied:
I do not want to hold the lock while calling callback..ok will fix it.

lpereira replied:
I know; locking interrupts for an indefinite amount of time isn't well advised.  There's not much you can do here, though.

One of the foreach functions that will leave the interrupt unlocked while calling the callback function is `net_tcp_foreach()`.  Using the same idea will not work, because it loops through an array of known size, not a linked list that can change size during iteration.

ramakrishnapallala replied:
yeah, agree. will fix it.

Commit:
kernel: Deprecate k_call_stacks_analyze() API

Deprecated k_call_stacks_analyze() API as it is only
dumping (printing) the statically defined main, idle,
work and ISR stacks.

Use k_thread_foreach() API which is a generic API
to iterate over threads.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
tests: kernel: profiling: Add test for k_thread_foreach API

Add a test case for k_thread_foreach API.

Replace deprecated k_call_stacks_analyze API with
k_thread_foreach for existing test cases.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
subsys: shell: Remove deprcated k_call_stacks_analyze API

Replace deprecated k_call_stacks_analyze() API with
k_thread_foreach() API.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
subsys: bluetooth: Remove deprcated k_call_stacks_analyze API

Replace deprecated k_call_stacks_analyze() API with
k_thread_foreach() API.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
samples: boards: Remove deprcated k_call_stacks_analyze API

Replace deprecated k_call_stacks_analyze() API with
k_thread_foreach() API.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@lpereira @pizi-nordic @andrewboie updated PR, please have a look.

pizi-nordic commented on include/kernel.h:
```diff
@@ -549,8 +551,26 @@ enum execution_context_types {
+ * @param user_cb Pointer to the user callback function.
+ * @param user_data Pointer to user data.
+ *
+ * @note CONFIG_THREAD_MONITOR must be set for this function to
```
The k_thread_foreach() does not produce any output, so this comment is not 100% correct.

pizi-nordic commented on kernel/thread.c:
```diff
@@ -36,6 +36,29 @@ extern struct _static_thread_data _static_thread_data_list_end[];
+	}
+}
+#else
+void k_thread_foreach(k_thread_user_cb_t user_cb, void *user_data) { }
```
The k_thread_foreach() might be used for something else than thread state inspection. However it silently fails when CONFIG_THREAD_MONITOR is not set. IMHO we should either make this function always available or add assertion informing user that this function is just a stub when CONFIG_THREAD_MONITOR  is not set.

ramakrishnapallala commented:
@pizi-nordic @lpereira updated PR, please have look. Thanks.

lpereira commented:
The race condition is still present in the last update.

Commit:
samples: subsys: debug: sysview: Adapt k_thread_foreach API

Use k_thread_foreach() API instead of iterating over the
_kernel.threads list.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@lpereira I tried my best avoid calling the user_cb with lock held. Looks like we can not do that due to risk of _kernel.threads being modified.

I have updated the PR with lock applied to the whole block. After looking at the sysview sample it made more sense to have irq_lock applied to whole block.

Please have relook into PR now.

ramakrishnapallala commented:
@lpereira please have a look at the updated PR.

lpereira reviewed (approved):
LGTM.

An improvement (for another patch) would be implementing k_call_stacks_analyze() in terms of k_thread_foreach().

ramakrishnapallala commented:
@andrewboie @pizi-nordic can you review the updated the PR.
